### PR TITLE
Adding t-out-of-N threshold access-structures to `drlwe` package

### PR DESCRIPTION
**Pull request by ChristianMct**

### CODE REVIEW

File: ring/ring_operations.go [hunk 1]
```diff
@@ -396,6 +414,16 @@ func (r *Ring) MulScalarBigintLvl(level int, p1 *Poly, scalar *big.Int, p2 *Poly
 	}
 }
 
+// EvalPolMontgomeryScalarNTT evaluate the polynomial pol at pk and writes the result in p3
+func (r *Ring) EvalPolMontgomeryScalarNTT(pol []*Poly, pk uint64, p3 *Poly) {
```
Comment (Pro7ech) [inline ring/ring_operations.go:418]: Rename to EvalPolyScalarMontgomery if that's the scalar that is supposed to be in Montgomery

File: ring/ring_operations.go [hunk 2]
```diff
@@ -396,6 +414,16 @@ func (r *Ring) MulScalarBigintLvl(level int, p1 *Poly, scalar *big.Int, p2 *Poly
...
+	p3.Copy(pol[len(pol)-1])
+	for i := len(pol) - 1; i > 0; i-- {
+		r.MulScalar(p3, pk, p3)
+		r.AddNoMod(p3, pol[i-1], p3)
```
Comment (Pro7ech) [inline ring/ring_operations.go:422]: Be careful of uint64 overflow, there is no check

File: ring/ring_operations.go [hunk 3]
```diff
@@ -396,6 +414,16 @@ func (r *Ring) MulScalarBigintLvl(level int, p1 *Poly, scalar *big.Int, p2 *Poly
 	}
 }
 
+// EvalPolMontgomeryScalarNTT evaluate the polynomial pol at pk and writes the result in p3
+func (r *Ring) EvalPolMontgomeryScalarNTT(pol []*Poly, pk uint64, p3 *Poly) {
+	p3.Copy(pol[len(pol)-1])
+	for i := len(pol) - 1; i > 0; i-- {
+		r.MulScalar(p3, pk, p3)
```
Comment (Pro7ech) [inline ring/ring_operations.go:421]: It doesn't use Montgomery?

File: ring/ring_scalar.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+// MulRNSScalar multiplies s1 and s2 and stores the result in sout.
+func (r *Ring) MulRNSScalar(s1, s2, sout RNSScalar) {
+	for i, qi := range r.Modulus {
+		sout[i] = MRedConstant(s1[i], s2[i], qi, r.MredParams[i])
+	}
+}
+
+// InverseCRT computes the modular inverse of a scalar a expressed in a CRT decomposition.
+// The inversion is done in-place and assumes that a is in Montgomery form.
+func (r *Ring) InverseCRT(a RNSScalar) {
```
Comment (Pro7ech) [inline ring/ring_scalar.go:40]: I would just call it ModInv

File: rlwe/ringqp/ringqp.go [hunk 5]
```diff
@@ -160,6 +160,63 @@ func (r *Ring) NegLvl(levelQ, levelP int, p1, p2 Poly) {
 	}
 }
 
+// NewScalar creates a new Scalar value (i.e., a degree-0 polynomial) in the RingQP.
+func (r *Ring) NewScalar() ring.RNSScalar {
```
Comment (Pro7ech) [inline rlwe/ringqp/ringqp.go:164]: NewRNSScalar

File: rlwe/ringqp/ringqp.go [hunk 6]
```diff
@@ -160,6 +160,63 @@ func (r *Ring) NegLvl(levelQ, levelP int, p1, p2 Poly) {
...
+	modlen := len(r.RingQ.Modulus)
+	if r.RingP != nil {
+		modlen += len(r.RingP.Modulus)
+	}
+	return make(ring.RNSScalar, modlen)
+}
+
+// NewScalarFromUInt64 creates a new Scalar in the RingQP initialized with value v.
+func (r *Ring) NewScalarFromUInt64(v uint64) ring.RNSScalar {
```
Comment (Pro7ech) [inline rlwe/ringqp/ringqp.go:173]: NewRNSScalarFromUInt64

File: rlwe/ringqp/ringqp.go [hunk 7]
```diff
@@ -160,6 +160,63 @@ func (r *Ring) NegLvl(levelQ, levelP int, p1, p2 Poly) {
...
+	if r.RingQ != nil {
+		r.RingQ.MulRNSScalar(s1[:qlen], s2[:qlen], sout[:qlen])
+	}
+	if r.RingP != nil {
+		r.RingP.MulRNSScalar(s1[qlen:], s2[qlen:], sout[qlen:])
+	}
+}
+
+// EvalPolMontgomeryScalarNTT evaluate the polynomial pol at pt and writes the result in p3
+func (r *Ring) EvalPolMontgomeryScalarNTT(pol []Poly, pt uint64, p3 Poly) {
```
Comment (Pro7ech) [inline rlwe/ringqp/ringqp.go:208]: EvalPolyScalarMontgomeryNTT

File: rlwe/ringqp/ringqp.go [hunk 8]
```diff
@@ -287,6 +344,30 @@ func (r *Ring) MulCoeffsMontgomeryAndAddLvl(levelQ, levelP int, p1, p2, p3 Poly)
 	}
 }
 
+// MulScalarCRT multiplies p with a scalar value expressed in the CRT decomposition.
+// It asssumes the scalar decomposition to be in Montgomerry form.
+func (r *Ring) MulScalarCRT(p Poly, scalar []uint64, pOut Poly) {
```
Comment (Pro7ech) [inline rlwe/ringqp/ringqp.go:349]: MulRNSScalar

File: rlwe/ringqp/ringqp.go [hunk 9]
```diff
@@ -287,6 +344,30 @@ func (r *Ring) MulCoeffsMontgomeryAndAddLvl(levelQ, levelP int, p1, p2, p3 Poly)
...
+		r.RingQ.MulScalarRNSScalar(p.Q, scalarQ, pOut.Q)
+	}
+	if r.RingP != nil {
+		r.RingP.MulScalarRNSScalar(p.P, scalarP, pOut.P)
+	}
+}
+
+// InverseCRT computes the modular inverse of a scalar a expressed in a CRT decomposition.
+// The inversion is done in-place and assumes that a is in Montgomery form.
+func (r *Ring) InverseCRT(scalar []uint64) {
```
Comment (Pro7ech) [inline rlwe/ringqp/ringqp.go:361]: ModInv

File: rlwe/ringqp/ringqp.go [hunk 10]
```diff
@@ -287,6 +344,30 @@ func (r *Ring) MulCoeffsMontgomeryAndAddLvl(levelQ, levelP int, p1, p2, p3 Poly)
 	}
 }
 
+// MulScalarCRT multiplies p with a scalar value expressed in the CRT decomposition.
+// It asssumes the scalar decomposition to be in Montgomerry form.
+func (r *Ring) MulScalarCRT(p Poly, scalar []uint64, pOut Poly) {
+	scalarQ, scalarP := scalar[:len(r.RingQ.Modulus)], scalar[len(r.RingQ.Modulus):]
+	if r.RingQ != nil {
```
Comment (Pro7ech) [inline rlwe/ringqp/ringqp.go:351]: Is there any case where RingQ == nil?

File: rlwe/ringqp/ringqp.go (hunk 9 above)
Comment (ChristianMct) [reply rlwe/ringqp/ringqp.go:361]: Use `Inverse(scalar RNSScalar)`

File: rlwe/ringqp/ringqp.go (hunk 10 above)
Comment (ChristianMct) [reply rlwe/ringqp/ringqp.go:351]: + Change to `MulScalarRNS`

File: drlwe/README.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+#### 2.iii.a Collective Key-Switching
+The parties perform a re-encryption of the desired ciphertext(s) from being encrypted under the _ideal secret-key_ to being encrypted under the receiver's secret-key.
+There are two instantiations of the Collective Key-Switching protocol:
+- Collective Key-Switching (CKS), implemented as the `drlwe.CKSProtocol` interface, enables the parties to switch from their _ideal secret-key_ _s_ to another _ideal secret-key_ _s'_ when s' is collectively known to the parties. In the case where _s' = 0_, this is equivalent to a decryption protocol which can be used when the receiver is one of the input-parties. 
+- Collective Public-Key Switching (PCKS), implemented as the `drlwe.PCKSProtocol` interface, enables parties to switch from their _ideal secret-key_ _s_ to an arbitrary key _s'_ when provided with a public encryption-key for _s'_. Hence, this enables key-switching to a secret-key that is not known to the input parties, which further enables external receivers.
+
+While both protocol variants have slightly different local operation, their steps are the same and as follows:
```
Comment (Pro7ech) [inline drlwe/README.md:150]: operations

File: drlwe/README.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+In the peer-to-peer setting, it could be a public broadcast channel.
+We also assume that parties can communicate over private authenticated channels.
+
+Several protocols require the parties to have access to common uniformly random polynomials (CRP), which are sampled from a common random string (CRS).
+This CRS is implemented as an interface type `drlwe.CRS` that can be read from the parties as a part of the protocols (see below).
+The `drlwe.CRS` can be implemented by a `utils.KeyedPRNG` type for which all parties use the same key.
+
+### 1. Setup
+In this phase, the parties generate the various keys that are required by the Evaluation phase.
+Similarly to LSSS-based MPC protocols such as SPDZ, the setup phase does not depend on the input and can be pre-computed.
```
[the remark this answers is not in the document]
Comment (ChristianMct) [reply drlwe/README.md:58]: I think that this is already captured by the first sentence.

**Pull request merged**